File: 6_kyu/Color_Choice.jl
module Color
    export checkchoose

    function checkchoose(m::Number, n::Integer)::Integer
        if m ≤ 0 || n ≤ 0
            # No solution
            return -1
        end
        if m == 1
            # special case, no handle by array of faculties
            return 0
        end
        faculty = Array{BigInt}(n)
        faculty[1] = 1
        for idx=2:n faculty[idx] = faculty[idx-1]*idx end
        # condition for x: m = n! / (x! * (n - x)!)
        # ⇔ x! * (n - x)! = n! / m
        if faculty[n] % m > 0
            return -1
        end
        target = faculty[n] ÷ m
        for x = 1:n÷2+1
            if faculty[x]*faculty[n-x] == target
                return x
            end
        end
        # No solution found
        -1
    end

end

File: 6_kyu/Color_Choice.lua
local solution = {}

-- Calculates x so than n over x == m.
-- Returns the smaller x if more then one solution exists.
-- Returns -1 if no solution exists.
function solution.checkChoose(m, n)
  -- handle invalid argument
  m = tonumber(m)
  n = tonumber(n)
  assert(m>0 and math.floor(m)==m,"argument #1 must be positive integer")
  assert(n>0 and math.floor(n)==n,"argument #2 must be positive integer")
  -- calculate binomial coeffizent over = n over x
  -- https://en.wikipedia.org/wiki/Binomial_coefficient
  local x = 0
  local over = 1
  -- handle floating point errors by "0.5" limit
  while over+0.5 < m and x < n do
    x = x + 1
    -- it is (n over x)=(n over x-1) * (n-x+1)/x
    over = over * ((n-x+1)/x)
  end
  if math.abs(m-over) < 0.5 then
    return x
  else
    -- no solution found, -1 indicates an error
    return -1
  end
end

return solution
